agent/xds/listeners_ingress.go

```diff
@@ -154,21 +137,108 @@ func (s *ResourceGenerator) makeIngressGatewayListeners(address string, cfgSnap
 	return resources, nil
 }
 
-func resolveListenerSDSConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*structs.GatewayTLSSDSConfig, error) {
+func makeDownstreamTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.DownstreamTlsContext, error) {
+	var downstreamContext *envoy_tls_v3.DownstreamTlsContext
+
+	tlsContext, err := makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap, listenerCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if tlsContext != nil {
+		downstreamContext = &envoy_tls_v3.DownstreamTlsContext{
+			CommonTlsContext:         tlsContext,
+			RequireClientCertificate: &wrappers.BoolValue{Value: false},
+		}
+	}
+
+	return downstreamContext, nil
+}
+
+func makeCommonTLSContextFromSnapshotListenerConfig(cfgSnap *proxycfg.ConfigSnapshot, listenerCfg structs.IngressListener) (*envoy_tls_v3.CommonTlsContext, error) {
+	var tlsContext *envoy_tls_v3.CommonTlsContext
+
+	// Enable connect TLS if it is enabled at the Gateway or specific listener
+	// level.
+	gatewayTLSCfg := cfgSnap.IngressGateway.TLSConfig
+
+	// TODO: should it be possible to explicitly _disable_ TLS on a listener if it's
+	// enabled on the gateway?
+	connectTLSEnabled := gatewayTLSCfg.Enabled ||
+		(listenerCfg.TLS != nil && listenerCfg.TLS.Enabled)
```

Comment:
@nickethier @banks This was existing logic I didn't change, but wasn't sure if it should be possible to disable TLS on a listener when it's enabled in the gateway config - currently that doesn't appear possible, although validation doesn't appear to error if a listener TLS block sets `Enabled: false`

Reply:
No because `false` is the zero value here which means that if the user only specifies TLS at gateway level and doesn't set anything at listener level we should NOT assume they want to disable TLS on all listeners!

For this reason we only take `true` as an explicit signal here so the only way to have partial listner support is to opt-in on all listeners with no GW level. Feel free to update comments if this wasn't clear enough!